### DISCUSSION TITLE
サブクエリ内でGroupingを扱う実装方法について（１度のSQLで集計を２回行う）

### DISCUSSION THREAD

**Discussion by grass-field** [domaframework/doma#1004] [Q&A]
いつも大変お世話になっております。
Criteria API を使用した SQL の構築で以下の抽出を検討しております。
```
⓵詳細（地区・家族単位）で集計し、集計した値を算術演算子等でデータ加工
⓶「⓵」の結果を地区単位で集計
```
実際のSQLとしては下記の様になります。
（１度のSQLで複数回の「GROUP BY」を使用したいです）
```
SELECT
    Chiku,
    SUM(divColumn01),
    SUM(divColumn02)
FROM
    (SELECT
        Chiku,
        Family,
        div(sum(ColumX, 1000)  AS divColumn01,
        div(sum(ColumY, 10000) AS divColumn02
    FROM TableMeisai
    GROUP BY
        Chiku,
        Family
    ) AS TableX
GROUP BY 
    Chiku
```
サブクエリ内でGroupingの方法を調べましたが、のような実装方法についての記述を見つけられませんでした。
この場合は、「⓵」の集計結果をテーブルに吐いて、再集計するのが最善の方法になりますでしょうか。

１度のSQLで集計ができれば思い、お問い合わせさせて頂きました。
こちらに関しても、実現方法や対応策等ご教示頂けますと幸いです。

**Comment by nakamura-to** [comment]
Criteria APIではFROM句にサブクエリを指定する記法はサポートしていません。
タイプセーフにしたり簡潔な記法にしたりが難しいためです。
このようなケースではSQLテンプレートを使っていただくのが良いと思います。

**Reply by grass-field** [reply]
承知しました。
 SQL Templateで検討します。
迅速なご回答有難うございます！
